### PR TITLE
Review Data Input Pipeline guides

### PR DESCRIPTION
**Pull request by melissawm**
# Description

This PR includes a review of the recently-added Data Input Pipeline guides (follow-up to https://github.com/AI-Hypercomputer/maxtext/pull/2244)

Some specific changes, including touching some files outside of the Data Input Pipeline guides:
- Documents must always start with a H1 header, and the subsequent headers must be sequential (i.e. an error will be raised in the ReadTheDocs/Sphinx build if a H1 header is followed by a H3 header, skipping a H2). 
- I added a few links I thought were appropriate, but I will flag them below.
- To be able to appear in the documentation, all documents must appear exactly once in a Table of Contents. In this case, I added a "toctree" directive to the guides/data_input_pipeline.md file but it is hidden. This means it does not appear in the rendered documentation page, but generates subsections for the child documents. See the screenshot below:

Captura de imagem_20250901_122356

# Tests

I have built the documentation pages locally to verify that it works. 

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: docs/guides/data_input_pipeline/data_input_grain.md [hunk 1]
```diff
@@ -1,75 +1,83 @@
...
+
+## The recommended input pipeline for determinism and resilience!
+
+[Grain](https://google-grain.readthedocs.io/en/latest/) is a library for reading data for training and evaluating JAX models. It’s designed to be:
 * **Powerful**: Users can bring arbitrary Python transformations.
 * **Flexible**: Users can readily override Grain components for their needs.
 * **Deterministic**: Multiple runs of the same pipeline will produce the same outputs.
 * **Resilient to preemptions**: With minimal-sized checkpoints, users can resume the dataloader from the point at which it was preempted and produce the same output as if it was never preempted.
 * **Performant**: Achieved with multiprocessing with shared memory. Tested on multiple data modalities.
 * **With minimal dependencies**: Does not depend on ML frameworks (Tensorflow).
```
Comment (melissawm) [inline docs/guides/data_input_pipeline/data_input_grain.md:12]: Do we need to add this here, since we are linking to the Grain website which contains the same info?

File: docs/guides/data_input_grain.md [hunk 2]
```diff
@@ -1,75 +1,83 @@
...
+## Why is determinism important for a data input pipeline?
 
----
+Determinism in a data input pipeline means that the same input data always results in the same sequence of batches at each step. This is typically achieved by setting a fixed [shuffle seed](https://google-grain.readthedocs.io/en/stable/_autosummary/grain.load.html) during pipeline initialization. In an ideal scenario, where training runs uninterrupted, this determinism is straightforward (deterministic without preemption). However, real-world distributed training environments often face preemptions due to maintenance, hardware failures, or resource constraints. When a preempted training run resumes, the data input pipeline is re-initialized. If the same shuffle seed is used, the pipeline restarts from the beginning, potentially re-training the model on initial data. Conversely, a new seed produces a different batch sequence, making it difficult to track which data has been seen and how often each example is used for training. This lack of control can impact model performance and reproducibility.
```
Comment (melissawm) [inline docs/guides/data_input_grain.md]: I am not sure if this is the correct link, but my suggestion is to add some sort of reference to the "shuffle seed" expression so readers have a bit more context.

File: docs/guides/data_input_grain.md [hunk 3]
```diff
@@ -1,75 +1,83 @@
...
+## How Grain achieves determinism
 
-### How Grain achieves determinism
-Grain ensures determinism in data input pipelines by saving the pipeline's state, including dataset metadata and processed data indices, within a small JSON file in checkpoints. When a training run is resumed with the same dataset and shuffle seed, Grain restores the pipeline's exact state from the checkpoint. This enables fully deterministic, reproducible training that is resilient to disruptions.
+Grain ensures determinism in data input pipelines by saving the pipeline's state, including dataset metadata and processed data indices, within a small JSON file in checkpoints (see [the Grain DataLoader tutorial](https://google-grain.readthedocs.io/en/stable/tutorials/data_loader_tutorial.html#checkpointing) or [the Grain checkpoint documentation](https://google-grain.readthedocs.io/en/stable/grain.checkpoint.html)). When a training run is resumed with the same dataset and shuffle seed, Grain restores the pipeline's exact state from the checkpoint. This enables fully deterministic, reproducible training that is resilient to disruptions.
```
Comment (melissawm) [inline docs/guides/data_input_grain.md]: Please confirm if these links are appropriate here.

File: docs/guides/data_input_grain.md [hunk 4]
```diff
@@ -1,75 +1,83 @@
...
+## Using Grain
+
+(Actually, see https://google-grain.readthedocs.io/en/stable/tutorials/data_sources/index.html)
+
+1. Grain currently supports two data formats: [ArrayRecord](https://google-grain.readthedocs.io/en/stable/tutorials/data_sources/arrayrecord_data_source_tutorial.html) (random access) and [Parquet](https://google-grain.readthedocs.io/en/stable/tutorials/data_sources/parquet_dataset_tutorial.html) (partial random-access through row groups). Only the ArrayRecord format supports the global shuffle mentioned above. For converting a dataset into ArrayRecord, see [Apache Beam Integration for ArrayRecord](https://github.com/google/array_record/tree/main/beam). Additionally, other random access data sources can be supported via a custom [data source](https://google-grain.readthedocs.io/en/latest/data_sources.html) class.
```
Comment (melissawm) [inline docs/guides/data_input_grain.md]: I do see more data formats in https://google-grain.readthedocs.io/en/stable/tutorials/data_sources/index.html, so I'm wondering if this is still correct that Grain only supports these two data formats.

File: docs/guides/data_input_grain.md [hunk 5]
```diff
@@ -1,75 +1,83 @@
...
+2. When the dataset is hosted on a Cloud Storage bucket, Grain can read it through
+   [Cloud Storage FUSE](https://cloud.google.com/storage/docs/gcs-fuse). The installation of Cloud Storage FUSE is included in [setup.sh](https://github.com/google/maxtext/blob/main/setup.sh). The user then needs to mount the Cloud Storage bucket to a local path for each worker, using the script [setup_gcsfuse.sh](https://github.com/google/maxtext/blob/main/setup_gcsfuse.sh). The script configures some parameters for the mount.
+
+   ```shell
+   bash setup_gcsfuse.sh \
+   DATASET_GCS_BUCKET=$BUCKET_NAME \
+   MOUNT_PATH=$MOUNT_PATH \
+   [FILE_PATH=$MOUNT_PATH/my_dataset]
+   ```
```
Comment (melissawm) [inline docs/guides/data_input_grain.md]: I added a shell language marker here so we get some syntax highlighting in the html output; let me know if you want it or not. Here's a screenshot of how this looks like:

Captura de imagem_20250901_124730

File: docs/guides/data_input_grain.md [hunk 6]
```diff
@@ -1,75 +1,83 @@
...
+
+   Note that `FILE_PATH` is optional; when provided, the script runs `ls -R` for pre-filling the metadata cache
+   (see ["Performance tuning best practices" on the Google Cloud documentation](https://cloud.google.com/storage/docs/cloud-storage-fuse/performance#improve-first-time-reads)).
+
+3. Set `dataset_type=grain`, `grain_file_type={arrayrecord|parquet}`, `grain_train_files` **(where?)** to match the file pattern on the mounted local path.
```
Comment (melissawm) [inline docs/guides/data_input_grain.md]: I believe these options should be set in the base.yml file or the command line, but wanted to confirm.

File: docs/guides/data_input_pipeline.md [hunk 7]
```diff
@@ -21,9 +21,9 @@ Currently MaxText has three data input pipelines:
...
 | **[Hugging Face](data_input_hf.md)** | datasets in [Hugging Face Hub](https://huggingface.co/datasets)<br>local/Cloud Storage datasets in json, parquet, arrow, csv, txt (sequential access) | no download needed, convenience; <br>multiple formats | limit scalability using the Hugging Face Hub (no limit using Cloud Storage); <br>non-deterministic with preemption<br>(deterministic without preemption)<br> |
-| **[TFDS](data_input_hf.md)** | TFRecord (sequential access), available through [Tensorflow Datasets](https://www.tensorflow.org/datasets/catalog/overview) | performant | only supports TFRecords; <br>non-deterministic with preemption<br>(deterministic without preemption) |
+| **[TFDS](data_input_tfds.md)** | TFRecord (sequential access), available through [Tensorflow Datasets](https://www.tensorflow.org/datasets/catalog/overview) | performant | only supports TFRecords; <br>non-deterministic with preemption<br>(deterministic without preemption) |
```
Comment (melissawm) [inline docs/guides/data_input_pipeline.md]: Please confirm if this link should be used here, but it seems like it was a typo.

File: docs/guides/data_input_pipeline.md [hunk 8]
```diff
@@ -37,11 +37,20 @@ The approaches to solve these challenges depend on whether your dataset supports
...
+* **Uneven completion**: Data indices are distributed evenly among hosts. Without packing, the data imbalance between hosts will be at most one batch. To handle the final steps where some hosts run out of data, you can enable the `generate_padding_example` flag **(where?)**. This directs hosts to generate empty "padding" batches until the training or evaluation steps are met. 
+
+```{note}
+When sequence packing is enabled, the difference in the number of packed examples per host can be larger. The `generate_padding_example` flag still solves this. However, as more hosts begin generating padding, you will observe a decrease in `total_weights` and a slower change in the training loss. If all hosts exhaust their data before the target step count is reached, both `total_weights` and loss will drop to 0.
+```
```
Comment (melissawm) [inline docs/guides/data_input_pipeline.md]: This syntax allows you to have nice "admonition"-style notes, here's a screenshot:

Captura de imagem_20250901_125716

Let me know if you want to use this here (or elsewhere in the docs)

File: docs/guides/data_input_pipeline.md [hunk 9]
```diff
@@ -37,11 +37,20 @@ The approaches to solve these challenges depend on whether your dataset supports
...
 * **Uneven completion**: Similar to random-access datasets, you can use the `generate_padding_example` flag to handle hosts that finish their file shards early (currently only supported in Hugging Face pipeline, not available in TFDS pipeline). 
 
+```{toctree}
+:hidden: true
 
-
+data_input_grain
+data_input_hf
+data_input_tfds
+```
```
Comment (melissawm) [inline docs/guides/data_input_pipeline.md]: This is the hidden toctree I mentioned above.

File: docs/guides/run_maxtext_via_xpk.md [hunk 10]
```diff
@@ -14,12 +14,12 @@
  limitations under the License.
 -->
 
-Running MaxText at Scale with XPK
-=================================
+(run-xpk)=
```
Comment (melissawm) [inline docs/guides/run_maxtext_via_xpk.md:17]: If this is removed, one of the internal links fails so we can't build the docs.

File: docs/guides/run_maxtext_via_xpk.md [hunk 11]
```diff
@@ -39,8 +39,7 @@ XPK abstracts away the complexity of cluster management and job submission, hand
 
 * * * * *
 
-1\. Prerequisites
------------------
+## 1. Prerequisites
```
Comment (melissawm) [inline docs/guides/run_maxtext_via_xpk.md]: This file was mixing rst and markdown-style syntax, so I just fixed the syntax for now.

File: docs/tutorials.md [hunk 12]
```diff
@@ -41,4 +41,5 @@ You can also find other examples in the [MaxText repository](https://github.com/
 tutorials/first_run.md
 tutorials/full_finetuning.md
 tutorials/run_llama2.md
+tutorials/grpo.md
```
Comment (melissawm) [inline docs/tutorials.md]: Please confirm if this is the correct place to add this new doc. I could not find a PR or discussion for this, so I am guessing it is a tutorial.

File: docs/guides/data_input_pipeline.md [hunk 13]
```diff
@@ -37,11 +37,20 @@ The approaches to solve these challenges depend on whether your dataset supports
...
-* **Uneven completion**: Data indices are distributed evenly among hosts. Without packing, the data imbalance between hosts will be at most one batch. To handle the final steps where some hosts run out of data, you can enable the `generate_padding_example` flag. This directs hosts to generate empty "padding" batches until the training or evaluation steps are met. **Note**: When sequence packing is enabled, the difference in the number of packed examples per host can be larger. The `generate_padding_example` flag still solves this. However, as more hosts begin generating padding, you will observe a decrease in total_weights and a slower change in the training loss. If all hosts exhaust their data before the target step count is reached, both total_weights and loss will drop to 0.
+* **Uneven completion**: Data indices are distributed evenly among hosts. Without packing, the data imbalance between hosts will be at most one batch. To handle the final steps where some hosts run out of data, you can enable the `generate_padding_example` flag **(where?)**. This directs hosts to generate empty "padding" batches until the training or evaluation steps are met.
```
Comment (melissawm) [inline docs/guides/data_input_pipeline.md]: Again, I believe this flag should be set in base.yml or the command line, but just want to confirm.

**Review by melissawm — COMMENTED** [review]
I left some comments below.

File: docs/guides/data_input_pipeline/data_input_grain.md (hunk 1 above)
Comment (aireenmei) [reply docs/guides/data_input_pipeline/data_input_grain.md:12]: I still prefer to have them here as well, since the user may not click the link. We want to increase visibility of these features.

File: docs/guides/data_input_grain.md (hunk 2 above)
Comment (aireenmei) [reply docs/guides/data_input_grain.md]: Thanks for checking. I think this link (https://google-grain.readthedocs.io/en/stable/grain.dataset.html#grain.MapDataset.shuffle) fits the context better, could you change it to this?

File: docs/guides/data_input_grain.md (hunk 6 above)
Comment (aireenmei) [reply docs/guides/data_input_grain.md]: They can be set in base.yml, or passed through command line (example command in the following paragraph). We can point this out either here or when we give example command

File: docs/guides/data_input_grain.md (hunk 3 above)
Comment (aireenmei) [reply docs/guides/data_input_grain.md]: The links are still appropriate. I think can change the description to "checkpointing section in Grain DataLoader tutorial" and "grain.checkpoint module" to be more accurate

File: docs/guides/data_input_pipeline.md (hunk 7 above)
Comment (aireenmei) [reply docs/guides/data_input_pipeline.md]: Yes, the link should be data_input_tfds.md. Thanks for catching it!

File: docs/guides/data_input_pipeline.md (hunk 13 above)
Comment (aireenmei) [reply docs/guides/data_input_pipeline.md]: yes, that's correct

File: docs/guides/data_input_pipeline.md (hunk 8 above)
Comment (aireenmei) [reply docs/guides/data_input_pipeline.md]: Yes, this looks nice

File: docs/guides/data_input_pipeline.md (hunk 9 above)
Comment (aireenmei) [reply docs/guides/data_input_pipeline.md]: Thanks for adding.

**Review by aireenmei — COMMENTED** [review]
Thanks for reviewing!

File: docs/guides/data_input_grain.md (hunk 4 above)
Comment (aireenmei) [reply docs/guides/data_input_grain.md]: You are right, thanks for checking. How about we change it to - Grain supports multiple [data sources](https://google-grain.readthedocs.io/en/stable/tutorials/data_sources/index.html). In the Grain pipeline in MaxText, we demonstrate the usage of two formats:  [ArrayRecord](https://google-grain.readthedocs.io/en/stable/tutorials/data_sources/arrayrecord_data_source_tutorial.html) (random access) and [Parquet](https://google-grain.readthedocs.io/en/stable/tutorials/data_sources/parquet_dataset_tutorial.html)

**Comment by melissawm** [comment]
This is ready for another round of review, thanks @aireenmei !

**Review by aireenmei — APPROVED** [review]
LGTM, thanks!

**Comment by melissawm** [comment]
Hi all, let me know if I can follow-up on this. Thanks

File: docs/guides/data_input_pipeline/data_input_tfds.md [hunk 14]
```diff
@@ -1,11 +1,14 @@
...
-```sh
-bash tools/data_generation/download_dataset.sh ${GCS_PROJECT} ${GCS_BUCKET_NAME}
+
+```shell
+bash download_dataset.sh {GCS_PROJECT} {GCS_BUCKET_NAME}
+```
+
+2. In [`src/MaxText/configs/base.yml`](https://github.com/AI-Hypercomputer/maxtext/blob/main/src/MaxText/configs/base.yml) or through command line, set the following parameters:
+
 ```
```
Comment (aireenmei) [inline docs/guides/data_input_pipeline/data_input_tfds.md]: Should this block be yaml?

Comment (melissawm) [inline docs/guides/data_input_pipeline/data_input_tfds.md]:
``` suggestion
```yaml
````

Comment (melissawm) [reply docs/guides/data_input_pipeline/data_input_tfds.md]: Updated, thanks!

File: docs/guides/data_input_pipeline.md [hunk 15]
```diff
@@ -36,7 +36,12 @@ The approaches to solve these challenges depend on whether your dataset supports
...
+
+* **Uneven completion**: Data indices are distributed evenly among hosts. Without packing, the data imbalance between hosts will be at most one batch. To handle the final steps where some hosts run out of data, you can enable the `generate_padding_batch_train`/`generate_padding_batch_eval` flag in `src/MaxText/config/base.yml` or through command line arguments. This directs hosts to generate empty "padding" batches until the training or evaluation steps are met.
+
+```{note}
+When sequence packing is enabled, the difference in the number of packed examples per host can be larger. The `generate_padding_batch_train`/`generate_padding_batch_eval` flag still solves this. However, as more hosts begin generating padding, you will observe a decrease in `total_weights` and a slower change in the training loss. If all hosts exhaust their data before the target step count is reached, both `total_weights` and loss will drop to 0.
```
Comment (jacoguzo) [inline docs/guides/data_input_pipeline.md]: can this be formatted as a paragraph rather than a single line?

Comment (melissawm) [reply docs/guides/data_input_pipeline.md]:
```suggestion
When sequence packing is enabled, the difference in the number of packed examples per host can be larger. The `generate_padding_batch_train`/`generate_padding_batch_eval` flag still solves this.

However, as more hosts begin generating padding, you will observe a decrease in `total_weights` and a slower change in the training loss.

If all hosts exhaust their data before the target step count is reached, both `total_weights` and loss will drop to 0.
```

**Review by jacoguzo — APPROVED** [review]

**Comment by hengtaoguo** [comment]
Hi @melissawm could you squash into one commit and sync to head?

**Comment by melissawm** [comment]
This is now rebased and squashed. Thanks!

**Comment by aireenmei** [comment]
Not sure why this PR is still open, I submitted the corresponding internal cl yesterday (https://github.com/AI-Hypercomputer/maxtext/commit/af29099b2209643e31619fd842cf646d82ca2b82)

**Comment by melissawm** [comment]
I see the changes on main @aireenmei , so this can be safely closed. I'm not sure if you need an extra step to do this internally so I'll leave it up to you. Thanks for checking!

**Comment by shralex** [comment]
Closing since changes are in main according to previous comment
